File: spotify-remote-queue/src/main/java/com/bryan/spotifyremotequeue/service/spotify/response/SearchResponse.java
package com.bryan.spotifyremotequeue.service.spotify.response;

import com.bryan.spotifyremotequeue.service.spotify.response.common.*;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.*;

@JsonIgnoreProperties
@Data
public class SearchResponse {

    private Tracks tracks;

    private Artists artists;

    private Albums albums;

    private Playlists playlists;

    private Shows shows;

    private Episodes episodes;

    private Audiobooks audiobooks;
}

@JsonIgnoreProperties
@Data
class Tracks {
    private String href;

    private int limit;

    private String next;

    private int offset;

    private String previous;

    private int total;

    private TrackObject[] items;
}

@JsonIgnoreProperties
@Data
class Artists {
    private String href;

    private int limit;

    private String next;

    private int offset;

    private String previous;

    private int total;

    private ArtistsArtist[] items;
}

@JsonIgnoreProperties
@Data
class ArtistsArtist {
    private ExternalUrls external_urls;

    private Followers followers;

    private String[] genres;

    private String href;

    private String id;

    private Image[] images;

    private String name;

    private int popularity;

    private String type;

    private String uri;
}

@JsonIgnoreProperties
@Data
class Albums {

    private String href;

    private int limit;

    private String next;

    private int offset;

    private String previous;

    private int total;

    private AlbumsAlbum[] items;
}

@JsonIgnoreProperties
@Data
class AlbumsAlbum {
    @JsonProperty("album_type")
    private String album_type;
    @JsonProperty("total_tracks")
    private int total_tracks;

    private String[] available_markets;

    private ExternalUrls external_urls;

    private String href;

    private String id;

    private Image[] images;

    private String name;
    @JsonProperty("release_date")
    private String release_date;
    @JsonProperty("release_date_precision")
    private String release_date_precision;

    private Restrictions restrictions;

    private String type;

    private String uri;

    private Copyright[] copyrights;

    private ExternalIds external_ids;

    private String[] genres;

    private String label;

    private int popularity;
    @JsonProperty("album_group")
    private String album_group;

    private AlbumsAlbumArtist[] artists;


}

@JsonIgnoreProperties
@Data
class AlbumsAlbumArtist {
    private ExternalUrls external_urls;

    private String href;

    private String id;

    private String name;

    private String type;

    private String uri;
}

@JsonIgnoreProperties
@Data
class Playlists {
    private String href;

    private int limit;

    private String next;

    private int offset;

    private String previous;

    private int total;

    private PlaylistsPlaylist[] items;
}

@JsonIgnoreProperties
@Data
class PlaylistsPlaylist {
    private boolean collaborative;

    private String description;

    private ExternalUrls external_urls;

    private String href;

    private String id;

    private Image[] images;

    private String name;

    private PlaylistsPlaylistOwner owner;

    @JsonProperty("public")
    private boolean isPublic;
    @JsonProperty("snapshot_id")
    private String snapshot_id;

    private PlaylistsPlaylistTrack tracks;

    private String type;

    private String uri;
}

@JsonIgnoreProperties
@Data
class PlaylistsPlaylistOwner {
    private ExternalUrls external_urls;

    private Followers followers;

    private String href;

    private String id;

    private String type;

    private String uri;
    @JsonProperty("display_name")
    private String display_name;
}

@JsonIgnoreProperties
@Data
class PlaylistsPlaylistTrack {
    private String href;

    private int total;
}

@JsonIgnoreProperties
@Data
class Shows {
    private String href;

    private int limit;

    private String next;

    private int offset;

    private String previous;

    private int total;

    private ShowsShow[] items;


}

@JsonIgnoreProperties
@Data
class ShowsShow {
    private String[] available_markets;

    private Copyright[] copyrights;

    private String description;
    @JsonProperty("html_description")
    private String html_description;

    private boolean explicit;

    private ExternalUrls external_urls;

    private String href;

    private String id;

    private Image[] images;
    @JsonProperty("is_externally_hosted")
    private boolean is_externally_hosted;

    private String[] languages;
    @JsonProperty("media_type")
    private String media_type;

    private String name;

    private String publisher;

    private String type;

    private String uri;
    @JsonProperty("total_episodes")
    private int total_episodes;
}

@JsonIgnoreProperties
@Data
class Episodes {
    private String href;

    private int limit;

    private String next;

    private int offset;

    private String previous;

    private int total;

    private EpisodesEpisode[] items;
}

@JsonIgnoreProperties
@Data
class EpisodesEpisode {
    @JsonProperty("audio_preview_url")
    private String audio_preview_url;

    private String description;
    @JsonProperty("html_description")
    private String html_description;
    @JsonProperty("duration_ms")
    private long duration_ms;

    private boolean explicit;

    private ExternalUrls external_urls;

    private String href;

    private String id;

    private Image[] images;
    @JsonProperty("is_externally_hosted")
    private boolean is_externally_hosted;
    @JsonProperty("is_playable")
    private boolean is_playable;

    private String language;

    private String[] languages;

    private String name;
    @JsonProperty("release_date")
    private String release_date;
    @JsonProperty("release_date_precision")
    private String release_date_precision;

    private EpisodesEpisodeResumePoint resume_point;

    private String type;

    private String uri;

    private Restrictions restrictions;
}

@JsonIgnoreProperties
@Data
class EpisodesEpisodeResumePoint {
    @JsonProperty("fully_played")
    private boolean fully_played;
    @JsonProperty("resume_position_ms")
    private long resume_position_ms;
}

@JsonIgnoreProperties
@Data
class Audiobooks {
    private String href;

    private int limit;

    private String next;

    private int offset;

    private String previous;

    private int total;

    private AudiobooksAudiobook[] items;
}

@JsonIgnoreProperties
@Data
class AudiobooksAudiobook {
    private AudiobooksAudiobookAuthor[] authors;

    private String[] available_markets;

    private Copyright[] copyrights;

    private String description;
    @JsonProperty("html_description")
    private String html_description;

    private String edition;

    private boolean explicit;

    private ExternalUrls external_urls;

    private String href;

    private String id;

    private Image[] images;

    private String[] languages;
    @JsonProperty("media_type")
    private String media_type;

    private String name;

    private AudiobooksAudiobookNarrator[] narrators;

    private String publisher;

    private String type;

    private String uri;
    @JsonProperty("total_chapters")
    private int total_chapters;
}

@JsonIgnoreProperties
@Data
class AudiobooksAudiobookAuthor {
    private String name;
}

@JsonIgnoreProperties
@Data
class AudiobooksAudiobookNarrator {
    private String name;
}

File: spotify-remote-queue-frontend/src/types/SearchResponse.ts
export type SearchResponse = {
	tracks: Tracks;
	artists: Artists;
	albums: Albums;
	playlists: Playlists;
	shows: Shows;
	episodes: Episodes;
	audiobooks: Audiobooks;
};

type Tracks = {
	href: string;
	limit: number;
	next: string;
	offset: number;
	previous: string;
	total: number;
	items: TracksTrack[];
};

export type TracksTrack = {
	album: TracksTrackAlbum;
	artists: TracksTrackArtists[];
	available_markets: string[];
	disc_number: number;
	duration_ms: number;
	explicit: boolean;
	external_ids: ExternalIds;
	external_urls: ExternalUrls;
	href: string;
	id: string;
	is_playable: boolean;
	linked_from: object;
	restrictions: Restrictions;
	name: string;
	popularity: number;
	preview_url: string;
	track_number: number;
	type: string;
	uri: string;
	is_local: boolean;
};

type TracksTrackAlbum = {
	album_type: string;
	total_tracks: number;
	available_markets: string[];
	external_urls: ExternalUrls;
	href: string;
	id: string;
	images: Image[];
	name: string;
	release_date: string;
	restrictions: Restrictions;
	type: string;
	copyrights: Copyrights[];
	external_ids: ExternalIds;
	genres: string[];
	label: string;
	popularity: number;
	album_group: string;
	artists: TracksTrackAlbumArtist[];
};

type TracksTrackAlbumArtist = {
	external_urls: ExternalUrls;
	href: string;
	id: string;
	name: string;
	type: string;
	uri: string;
};

type TracksTrackArtists = {
	external_urls: ExternalUrls;
	followers: Followers;
	genres: string[];
	href: string;
	id: string;
	images: Image[];
	name: string;
	popularity: number;
	type: string;
	uri: string;
};

type Artists = {
	href: string;
	limit: number;
	next: string;
	offset: number;
	previous: string;
	total: number;
	items: ArtistsArtist[];
};

type ArtistsArtist = {
	external_urls: ExternalUrls;
	followers: Followers;
	genres: string[];
	href: string;
	id: string;
	images: Image[];
	name: string;
	popularity: number;
	type: string;
	uri: string;
};

type Albums = {
	href: string;
	limit: number;
	next: string;
	offset: number;
	previous: string;
	total: number;
	items: AlbumsAlbum[];
};

type AlbumsAlbum = {
	album_type: string;
	total_tracks: number;
	available_markets: string[];
	external_urls: ExternalUrls;
	href: string;
	id: string;
	images: Image[];
	name: string;
	release_date: string;
	release_date_precision: string;
	restrictions: Restrictions;
	type: string;
	uri: string;
	copyrights: Copyrights;
	external_ids: ExternalIds;
	genres: string[];
	label: string;
	popularity: number;
	album_group: string;
	artists: AlbumsAlbumArtist[];
};

type AlbumsAlbumArtist = {
	external_urls: ExternalUrls;
	href: string;
	id: string;
	name: string;
	type: string;
	uri: string;
};

type Playlists = {
	href: string;
	limit: number;
	next: string;
	offset: number;
	previous: string;
	total: number;
	items: PlaylistsPlaylist[];
};

type PlaylistsPlaylist = {
	collaborative: boolean;
	description: string;
	external_urls: ExternalUrls;
	href: string;
	id: string;
	images: Image[];
	name: string;
	owner: {
		external_urls: ExternalUrls;
		followers: Followers;
		href: string;
		id: string;
		type: string;
		uri: string;
		display_name: string;
	};
	public: boolean;
	snapshot_id: string;
	tracks: PlaylistsPlaylistTracks;
	type: string;
	uri: string;
};

type PlaylistsPlaylistTracks = {
	href: string;
	total: number;
};

type Shows = {
	href: string;
	limit: number;
	next: string;
	offset: number;
	previous: string;
	total: number;
	items: ShowsShow[];
};

type ShowsShow = {
	available_markets: string[];
	copyrights: Copyrights;
	description: string;
	html_description: string;
	explicit: boolean;
	external_urls: ExternalUrls;
	href: string;
	id: string;
	images: Image[];
	is_externally_hosted: boolean;
	languages: string[];
	media_type: string;
	name: string;
	publisher: string;
	type: string;
	uri: string;
	total_episodes: number;
};

type Episodes = {
	href: string;
	limit: number;
	next: string;
	offset: number;
	previous: string;
	total: number;
	items: EpisodesEpisode[];
};

type EpisodesEpisode = {
	audio_preview_url: string;
	description: string;
	html_description: string;
	duration_ms: number;
	explicit: boolean;
	external_urls: ExternalUrls;
	href: string;
	id: string;
	images: Image[];
	is_externally_hosted: string;
	is_playable: string;
	language: string;
	languages: string[];
	name: string;
	release_date: string;
	release_date_precision: string;
	resume_point: {
		fully_played: boolean;
		resume_position_ms: number;
	};
	type: string;
	uri: string;
	restrictions: Restrictions;
};

type Audiobooks = {
	href: string;
	limit: number;
	next: string;
	offset: number;
	previous: string;
	total: number;
	items: AudiobooksAudiobook[];
};

type AudiobooksAudiobook = {
	authors: {
		name: string;
	}[];
	available_markets: string[];
	copyrights: Copyrights;
	description: string;
	html_description: string;
	edition: string;
	explicit: boolean;
	external_urls: ExternalUrls;
	href: string;
	id: string;
	images: Image[];
	languages: string[];
	media_type: string;
	name: string;
	narrators: {
		name: string;
	}[];
	publisher: string;
	type: string;
	uri: string;
	total_chapters: number;
};

type ExternalUrls = {
	spotify: string;
};

type Image = {
	url: string;
	height: number;
	width: number;
};

type Restrictions = {
	reason: string;
};

type Copyrights = {
	text: string;
	type: string;
};

type ExternalIds = {
	isrc: string;
	ean: string;
	upc: string;
};

type Followers = {
	href: string;
	total: number;
};
